source/Irrlicht/CBAWMeshWriter.cpp

```diff
@@ -0,0 +1,269 @@
+
+		CBAWMeshWriter::~CBAWMeshWriter() {}
+
+		CBAWMeshWriter::CBAWMeshWriter()
+		{
+#ifdef _DEBUG
+			setDebugName("CBAWMeshWriter");
+#endif
+		}
+
+		template<>
+		void CBAWMeshWriter::exportAsBlob<ICPUMeshBuffer>(ICPUMeshBuffer* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			uint8_t* const data = (uint8_t*)std::malloc(m_headers[_headerIdx].blobSizeDecompr);
+			uint8_t* nearestFree = data;
+			std::memcpy(nearestFree, &_obj->getMaterial(), sizeof(video::SMaterial));
+			nearestFree += sizeof(video::SMaterial);
+			std::memcpy(nearestFree, &_obj->getBoundingBox(), sizeof(core::aabbox3df));
+			nearestFree += sizeof(core::aabbox3df);
+			*((uint64_t *)nearestFree) = reinterpret_cast<uint64_t>(_obj->getMeshDataAndFormat());
+			nearestFree += sizeof(uint64_t);
+			*((video::E_INDEX_TYPE *)nearestFree) = _obj->getIndexType();
+			nearestFree += sizeof(video::E_INDEX_TYPE);
+			*((int32_t *)nearestFree) = _obj->getBaseVertex();
+			nearestFree += sizeof(int32_t);
+			*((uint64_t *)nearestFree) = _obj->getIndexCount();
+			nearestFree += sizeof(uint64_t);
+			*((size_t *)nearestFree) = _obj->getIndexBufferOffset();
+			nearestFree += sizeof(size_t);
+			*((size_t *)nearestFree) = _obj->getInstanceCount();
+			nearestFree += sizeof(size_t);
+			*((uint32_t *)nearestFree) = _obj->getBaseInstance();
+			nearestFree += sizeof(uint32_t);
+			*((E_PRIMITIVE_TYPE *)nearestFree) = _obj->getPrimitiveType();
+			nearestFree += sizeof(E_PRIMITIVE_TYPE);
+			*((E_VERTEX_ATTRIBUTE_ID *)nearestFree) = _obj->getPositionAttributeIx();
+			nearestFree += sizeof(E_VERTEX_ATTRIBUTE_ID);
+
+			core::XXHash_256(data, m_headers[_headerIdx].blobSizeDecompr, m_headers[_headerIdx].blobHash);
+			_file->write(data, m_headers[_headerIdx].blobSizeDecompr);
+			calcAndPushNextOffset(!_headerIdx ? 0 : m_headers[_headerIdx-1].blobSizeDecompr);
+
+			std::free(data);
+		}
+		template<>
+		void CBAWMeshWriter::exportAsBlob<video::IVirtualTexture>(video::IVirtualTexture* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			video::ITexture* tex;
+			if (!(tex = dynamic_cast<video::ITexture*>(_obj)))
+				return;
+			//! @todo @bug here we get absolute path and we want relative
+			const io::path path = tex->getName().getInternalName();
+			const uint32_t len = std::strlen(path.c_str()) + 1;
+
+			m_headers[_headerIdx].blobSize = m_headers[_headerIdx].blobSizeDecompr = len;
+			core::XXHash_256(path.c_str(), len, m_headers[_headerIdx].blobHash);
+			calcAndPushNextOffset(!_headerIdx ? 0 : m_headers[_headerIdx-1].blobSizeDecompr);
+
+			_file->write(path.c_str(), len);
+		}
+		template<>
+		void CBAWMeshWriter::exportAsBlob<IMeshDataFormatDesc<core::ICPUBuffer> >(IMeshDataFormatDesc<core::ICPUBuffer>* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			uint8_t* const data = (uint8_t*)std::malloc(m_headers[_headerIdx].blobSizeDecompr);
+
+			E_COMPONENTS_PER_ATTRIBUTE* cpa = (E_COMPONENTS_PER_ATTRIBUTE*)data;
```

Comment:
All this stuff is nice, but not very maintanable.. maybe make a "struct MeshDataFormatDescBlobV1 : public Blob" which will construct itself from an IMeshDataFormatDesc<ICPUBuffer>*

And also use the stack instead of heap! 
Making our exporter faster!